RFC-0001-native-spi.md

```diff
@@ -0,0 +1,574 @@
+
+HTTP endpoints related to tasks are registered to Proxygen in `TaskResource.cpp`. Important endpoints implemented include:
+
+
+
+* POST: v1/task: This processes a `TaskUpdateRequest`
+* GET: v1/task: This returns a serialized `TaskInfo` (used for comprehensive metrics, may be reported less frequently)
+* GET: v1/task/status: This returns a serialized `TaskStatus` (used for query progress tracking, must be reported frequently)
+
+Other HTTP endpoints include:
+
+
+
+* POST: v1/memory
+    * This just reports memory, but no assignments are adjusted unlike in Java workers.
+* GET: v1/info
+* GET: v1/status
+
+The request/response flow of Prestissimo is identical to Java workers. The tasks, or new splits, are registered via `TaskUpdateRequest`. Resource utilization and query progress are sent to the coordinator via task endpoints.
+
+![Prestissimo build](RFC-0002/build.png)
+
+
+
+## Current Needs
+
+
+### Fail fast 
+
+
+#### Fail fast for queries that aren’t supported 
+
+Prestissimo does not use the Plugin interface, and so lacks knowledge of built-in functions, built-in types, operators, or connectors. If one attempts to reference any of these custom additions, then the query will fail during execution. While in an under-utilized interactive cluster this may be barely noticeable, some clusters often experience heavy queueing. Even if a cluster experiences only periodic queueing, failing during execution may make the queueing worse as queries which are destined to fail must first be queued and planned for execution before the query fails. We should allow Prestissimo clusters to fail fast queries which will never succeed during execution, similar to Java clusters.
+
+
+#### DDL type support should be consistent with execution 
+
+Because type support is different between the Java SPI and what is supported in Prestissimo, it’s possible to create views and tables that utilize unsupported types in Prestissimo. For the best user experience, the support should be seamless between DDL and execution. Additionally, for developer ease, there should be just one SPI surface that one must implement to register the built-in types. Finally, it would be nice to implement a system procedure that validates that a type is supported for all tables in a schema, so that customers can quickly understand the support level of the underlying eval.
+
+
+#### Configuration properties 
+
+Presto validates which configuration properties end up being used by ensuring that all entries in the configuration correspond to a configuration file in the codebase. If one references a configuration that is unused then the Presto server will fail to start. In Java clusters, there are essentially three categories of configuration:
+
+
+
+1. Configuration that affects coordinators only
+2. Configuration that affects workers only
+3. Configuration that affects both
```

Comment:
Hi @tdcmeehan would it be possible to add some examples for the 3 categories?

Reply:
Done